lib/streamlit/commands/navigation.py

```diff
@@ -0,0 +1,189 @@
+    `st.Page`. `st.navigation` returns the current page which can be executed
+    using `Page.run()`.
+
+    When using `st.navigation`, the main script is executed on every app rerun,
+    with the current page executed in-line using `Page.run()`. The set of
+    available pages can be updated with each rerun for dynamic navigation. In
+    this mode, the `pages/` folder is ignored.
+
+    By default, `st.navigation` draws the available pages in the side
+    navigation. This behavior can be changed using the `position=` keyword
+    argument.
+
+    Parameters
+    ----------
+    pages: list[st.Page] or dict[str, list[st.Page]]
+        A list of `st.Page` objects or a dictionary where the keys are section
+        headers and the values are lists of `st.Page` objects.
+
+    position: str
+        The position of the navigation menu. Can be "sidebar" or "hidden".
+
+    Example
+    -------
+    >>> import streamlit as st
+    >>> from pages import page1, page2
+    >>>
+    >>>	pg = st.navigation([st.Page(page1), st.Page(page2)])
+    >>>
+    >>>	st.title("My Awesome App")
+    >>> pg.run()
+    """
+    ctx = get_script_run_ctx()
+    if not ctx:
+        return None
+
+    nav_sections = {"": pages} if isinstance(pages, list) else pages
+    page_list = pages_from_nav_sections(nav_sections)
+
+    if not page_list:
+        raise StreamlitAPIException(
+            "`st.navigation` must be called with at least one `st.Page`."
+        )
+
+    default_page = None
+    pagehash_to_pageinfo: dict[PageHash, PageInfo] = {}
+
+    # This nested loop keeps track of three things:
+    # 1. the default page
+    # 2. the pagehash to pageinfo mapping
+    for section_header in nav_sections:
+        for page in nav_sections[section_header]:
+            if page.default:
+                if default_page is not None:
+                    raise StreamlitAPIException(
+                        "Multiple Pages specified with `default=True`. "
+                        "At most one Page can be set to default."
+                    )
+                default_page = page
+
+            if isinstance(page._page, Path):
+                script_path = str(page._page)
+            else:
+                script_path = ""
+
+            script_hash = page._script_hash
+            if script_hash in pagehash_to_pageinfo:
+                raise StreamlitAPIException(
+                    f"Multiple Pages specified with title {page.title}. "
+                    "Page titles should be unique. If not specified, "
+                    "title is inferred from the file path or callable name."
+                )
+
+            pagehash_to_pageinfo[script_hash] = {
+                "page_script_hash": script_hash,
+                "page_name": page.title,
+                "icon": page.icon,
+                "script_path": script_path,
+                "url_pathname": page.title.replace(" ", "_"),
```

Comment:
nit: should we also lowercase the path name? Case-sensitive path names look a bit ugly. Also, it might make sense to put the logic into a shared utils method - e.g., `title_to_url_path` - since it's at least done one more time below. And its easy to otherwise lose track of this and only apply the change at one point.

Reply:
I think this is a good question, but I think it mimics a similar behavior to MPAv1 and encourage people to use a follow-up `url_path` keyword param to modify it to their liking.